[Detection Engine][Rule Suppression] Add Suppression to EQL Non-sequence based queries

WafaaNasr opened the pull request:
# Summary

- Address adding suppression to EQL rules https://github.com/elastic/security-team/issues/7773
- Milestone details https://github.com/elastic/security-team/issues/8432

## Checklist
- [x] Functional changes are hidden behind a feature flag. If not hidden, the PR explains why these changes are being implemented in a long-living feature branch. 
- [x] Functional changes are covered with a test plan and automated tests. [Test plan](https://github.com/elastic/security-team/pull/9155)
- [ ] Stability of new and changed tests is verified using the [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner) in both ESS and Serverless. By default, use 200 runs for ESS and 200 runs for Serverless.
    * Cypress ESS: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5686
    * Cypress Serverless: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5687
    * FTR ESS: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5688
    * FTR Serverless: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5689   
- [x] Comprehensive manual testing is done by two engineers: the PR author and one of the PR reviewers. Changes are tested in both ESS and Serverless.
- [x] Mapping changes are accompanied by a technical design document. It can be a GitHub issue or an RFC explaining the changes. The design document is shared with and approved by the appropriate teams and individual stakeholders.
- [ ] (OPTIONAL) OpenAPI specs changes include detailed descriptions and examples of usage and are ready to be released on https://docs.elastic.co/api-reference. NOTE: This is optional because at the moment we don't have yet any OpenAPI specs that would be fully "documented" and "GA-ready" for publishing on https://docs.elastic.co/api-reference.
- [x] Functional changes are communicated to the Docs team. A ticket is opened in https://github.com/elastic/security-docs using the [`Internal documentation request (Elastic employees)`](https://github.com/elastic/security-docs/issues/new?assignees=&labels=&projects=&template=docs-request-internal.yaml&title=%5BRequest%5D+) template. The following information is included: feature flags used, target ESS version, planned timing for ESS and Serverless releases.
- [x] Check if in timeline we can show the suppression count column when the user clicks on investigate on timeline for Eql suppressed Alerts (https://github.com/elastic/kibana/issues/180976)

## Related Issues
* Sub-PRs
    - Address EQL schema changes PR https://github.com/elastic/kibana/pull/176391
        - Adding Feature flag PR and updating the Frontend Part in Rule Create/Edit https://github.com/elastic/kibana/pull/176398
    - Adding Backend changes and FTR tests https://github.com/elastic/kibana/pull/176597
    - Fix Investigate in Timeline for the Suppressed Alerts https://github.com/elastic/kibana/pull/177839
    - Add Cypress e2e tests https://github.com/elastic/kibana/pull/177870
    - Disable EQL sequence suppression in the UI and fix Cypress `after` esArchive path https://github.com/elastic/kibana/pull/178531
- Docs Issue https://github.com/elastic/security-docs/issues/4977
- Test plan https://github.com/elastic/security-team/pull/9155

## Screenshots/recordings

### Non-Sequence Suppression

1. Rule creation, Suppression based on a single value

 https://github.com/elastic/kibana/assets/12671903/8d168bce-15d3-45c2-a5dc-238b3ac01626

2. Rule creation, Suppression based on an array of values
  
https://github.com/elastic/kibana/assets/12671903/0e3312a9-4eae-476b-9c1e-c68189bbaf95

3. Investigate In Timeline

https://github.com/elastic/kibana/assets/12671903/e10c8668-4d5b-4748-b8a1-678603b4a8a5


### Disabled Sequence Suppression

1. UI

https://github.com/elastic/kibana/assets/12671903/01faa649-ca8b-43e4-a398-42ab242e7a72

WafaaNasr committed:
add suppression to eql in the FE + adding FF

WafaaNasr committed:
add the return type

WafaaNasr committed:
fix type

WafaaNasr committed:
add initial UI tests

WafaaNasr committed:
[Detection Engine][Rule Suppression] Add suppression fields to EQL schema (#176391)

## Summary
- Address schema changes
- Add Tests in rule_converters to validate converting from/to snake case
to/from camel case

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-FE-FF-changes

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-FE-FF-changes

WafaaNasr commented:
/ci

WafaaNasr committed:
initial implementation

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

WafaaNasr committed:
add ftr tests for eql non-sequence query, missing warning cases

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
undo tsconfig

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

WafaaNasr committed:
adding is_alert_suppression_active util + tests

WafaaNasr committed:
fix conflict

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

WafaaNasr committed:
fix non-suppression scenario

WafaaNasr committed:
Merge branch 'eql-suppression-BE-FTR-changes' of https://github.com/WafaaNasr/kibana into eql-suppression-BE-FTR-changes

WafaaNasr committed:
remove console.log

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

WafaaNasr committed:
fix investigate in timeline until confirmation

WafaaNasr committed:
add warning for suppressed max signal message

WafaaNasr committed:
Merge branch 'eql-suppression-BE-FTR-changes' of https://github.com/WafaaNasr/kibana into eql-suppression-BE-FTR-changes

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

WafaaNasr committed:
add tips for sequence implementation

WafaaNasr committed:
Merge branch 'eql-suppression-BE-FTR-changes' of https://github.com/WafaaNasr/kibana into eql-suppression-BE-FTR-changes

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
[Detection Engine][Rule Suppression] Add suppression fields to EQL Non-sequence based queries- Create/Edit FE components (#176398)

## Summary

- Adding the Frontend part to enable suppression in EQL rules based on a
Feature flag
- Adding tests to:
     - useAlertSuppression hook to include Eql type
- rule_creation/helpers.test.ts to validate the Alert Suppression field
UI part in the Define step in Rule creation/edit (need to ask about
Group by)
     - Add validation for only using max 3 group by fields + unit tests

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

WafaaNasr committed:
remove extra spaces

WafaaNasr committed:
fix ff name

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-BE-FTR-changes

WafaaNasr committed:
fix tests

WafaaNasr committed:
Merge branch 'eql-suppression-BE-FTR-changes' of https://github.com/WafaaNasr/kibana into eql-suppression-BE-FTR-changes

WafaaNasr committed:
update schema snapshot for serverless

WafaaNasr committed:
fix checktypes

WafaaNasr committed:
separate the timeline changes in different pr

WafaaNasr committed:
reuse bulk_created_suppressed_alerts_in_memory in search_after_bulk_create_suppressed_alerts

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
add cypress e2e for eql suppression non-sequence

WafaaNasr committed:
rename

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
[Detection Engine][Rule Suppression] Eql suppression Backend implementation and FTR tests for Non-sequence queries (#176597)

## Summary

- Adding the backend part to enable suppression in EQL rules based on a
Feature flag
- Adding is_alert_suppression_active util
- Adding tests:
- ESS and Serverless [Non-sequence query suppression
](https://github.com/elastic/kibana/blob/9049b9abc29a9e46e6e4c3e712c42bcab13459cb/x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts)
- Unit tests for
[is_alert_suppression_active](https://github.com/elastic/kibana/blob/b8a370b8f0960b3a981807c7ac79ec6771be5933/x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/is_alert_suppression_active.test.ts)
- Handle Warning cases for max suppressed alerts if suppression is
active
- Reuse the `search_after_bulk_create_suppressed_alerts` in
`search_after_bulk_create_suppressed_alerts`
- TODO: add desc around FTR test for non-sequence based queries

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-cypress-non-seq

WafaaNasr committed:
add extra line

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
[Detection Engine][Rule Suppression]  Non-sequence  Eql suppression Cypress e2e tests (#177870)

## Summary

Adding UI-E2E tests for the Non-sequence Alerts generated by EQL
suppression
 - Serverless - Essentials
 - ESS - Basic
 - ESS - Platinum and Serverless - Complete

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr commented:
/ci

WafaaNasr committed:
[Detection Engine][Rule Suppression] Eql suppression for Non-sequence queries - Fix Investigate in Timelines (#177839)

## Summary


Correct the behavior of populating the EQL query in the query bar when
examining suppressed non-sequence alerts in the Timeline

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

WafaaNasr committed:
disable eql sequence suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-disable-sequence

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-disable-sequence

WafaaNasr committed:
add tests and todo for be exception

WafaaNasr committed:
Merge branch 'main' into security/eql-suppression

WafaaNasr committed:
Merge branch 'security/eql-suppression' into eql-suppression-disable-sequence

WafaaNasr committed:
fix after in cypress

WafaaNasr committed:
disable eql sequence suppression and ask the user to reset the suppression fields if exist

WafaaNasr committed:
add todo to ignore suppression for sequence query if passed

rylnd committed:
Use a slightly more legible feature flag name

This attempts to more clearly divide the base verb ("is enabled")
from the noun + adjectives ("alert suppression for non-sequence EQL
rule").

rylnd committed:
Replace validation-specific predicate with more general helper

The logic here is identical, and the existing function
(`isSuppressionRuleConfiguredWithGroupBy`), is both better-named and
more general.

rylnd committed:
Merge branch 'main' into security/eql-suppression

 Conflicts:
	x-pack/plugins/security_solution/common/experimental_features.ts
	x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/use_experimental_feature_fields_transform.ts
	x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.test.tsx
	x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.tsx
	x-pack/test/security_solution_api_integration/config/ess/config.base.ts
	x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts
	x-pack/test/security_solution_cypress/config.ts
	x-pack/test/security_solution_cypress/serverless_config.ts

rylnd commented:
/ci

rylnd committed:
Remove unused test file

This referenced modules that were removed previously in 6aa3dbe.

rylnd committed:
Update unload task invocations after API changed

The changes in 178724 made the simple string argument no longer
supported, here.

rylnd committed:
Fix our mocks in the useAlertSuppression hook

The use of `useIsExperimentalFeatureEnabled` was removed on `main`, but
we still have it (and thus need to mock it) on this branch.

rylnd commented:
/ci

rylnd committed:
Fix bad merge resolution

This key was deleted upstream.

rylnd committed:
refactors EQL execution to allow for a handling of sequence suppression

For now, this is just going to display a warning, but it will eventually
be where we handle sequence suppression itself.

This was a bit of a tricky refactor, but it should be the same. I was
able to get rid of some chained ternaries, and I found that we weren't
adding our suppression results to the rule's results (for which I've
added a TODO to investigate).

Next up: adding the warning.

rylnd committed:
Sequence EQL rules with suppression are marked as partial failure

This shouldn't be possible via the UI, and would only occur if a hopeful
API user creates/updates a rule to have both suppression and an EQL
sequence query. In that case, we warn them that the suppression
configuration will be ignored, but that the rule will otherwise execute
as normal.

I mainly made this a partial failure (and not simply a log message) for
ease of testing (since the status change will be returned by the rule
preview endpoint); if we want to switch this later that's easy enough.

rylnd commented:
/ci

rylnd committed:
Merge branch 'main' into security/eql-suppression

rylnd commented:
/ci

rylnd committed:
Simplify warning if using suppression with EQL sequences

The partial failure is automatically generated if a warning is present;
we don't need the extra call (I'm currently writing integration tests
that verify this).

rylnd committed:
Remove unnecessary addToSearchAfterReturn call

This already happens within `bulkCreateSuppressedAlertsInMemory`, so the
external call is not only unnecessary but incorrect.

I've added a test to verify that _something_ gets reported in the
suppression case, but it appears that none of those suppression fields
are recorded in the execution status itself.

rylnd committed:
Attempting to test the fact that we suppressed alerts

The BulkCreateResult has `suppressedAlertsCount: 9` in this scenario,
but I don't know if that makes it to the user in any way that we can
test.

rylnd commented:
/ci

rylnd committed:
style: give tests some breathing room

rylnd committed:
Slightly more comprehensive test assertion

Assert that we're recording both index and search time in our execution
when doing suppression.

rylnd commented:
/ci

rylnd committed:
Merge branch 'security/eql-suppression' into eql-suppression-disable-sequence

 Conflicts:
	x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx
	x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/eql/eql.ts
	x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule_suppression_ess_basic.cy.ts

rylnd committed:
Ignore complexity post-merge

It's official, we've now got way too much happening in this component!

rylnd committed:
Alternate implementation of disabling suppression for EQL sequences

At its core, this is driven by `useFormData` (as I think a lot of
complexity in this form could be):

* We use the current, updated value of `queryBar`, along with the
  `groupByFields`, to determine whether the suppression fields should be
  disabled. The core logic here is the same as before.
* We move the _validity_ of the groupByFields field into the form
  validations. Again, the logic here is the same, but it's now done a)
  only when the relevant fields change, and b) without the need for
  manually calling `validateFields` or `setFieldErrors`.

Other notes:
* Adds an `isEqlSequenceQuery` helper for explicitness.
* Removes callback hook from query bar, since we now get this info
  directly from the form.

rylnd committed:
Update copy as per docs team feedback

These were discussed in elastic/security-docs#4977

rylnd committed:
Fix translation interpolation

rylnd committed:
Fix translation key

These were duplicated since their inception, but CI was never run so it
wasn't caught until now.

rylnd committed:
Merge branch 'main' into security/eql-suppression

rylnd committed:
Fix unit tests for EQL query bar

Removes tests around deleted prop, fixes imports

rylnd committed:
Add some tests around added helper

rylnd committed:
Merge branch 'main' into security/eql-suppression

 Conflicts:
	x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.schema.yaml

rylnd committed:
Merge branch 'security/eql-suppression' into eql-suppression-disable-sequence

rylnd committed:
Fix outdated feature flag reference

This one snuck through the myriad merges/conflict resolutions on this
branch.

rylnd committed:
Remove assertion about upsell tooltip when suppression is disabled

See TODO for details.

rylnd committed:
Only show "disabled for EQL sequence" tooltip on suppression fields
input

This tooltip was previously wrapping several form fields (all of those
related to alert suppression, which:

1. Caused strange UX when attempting to hover disabled elements
  * Because many EUI disabled elements have `pointer-events: none`, you
    could only trigger the tooltip in certain areas of the form
  * I suspect there might've been some interaction with nested tooltips,
    where if you hovered the inner, the outer would also open and they
    would cancel each other out.
2. Caused existing cypress tests around suppression tooltips to fail,
   mainly due to the above.

For these reasons, I've limited the scope of this tooltip to only show
on hover of the alert suppression (group by) input, which is both more
consistent and much more robust, UX-wise.

The only downside here is that if you're on a basic license AND you type
a sequence query, you'll get one of the two tooltips depending on where
you hover, but that was already the case before this.

rylnd committed:
Revert "Remove assertion about upsell tooltip when suppression is disabled"

This reverts commit f974273fa1e49d9c4a64a024afb6ac93ee31b131.

See previous commit about our fixing of the tooltip.

rylnd committed:
Fix cypress assertion following copy change

rylnd committed:
Merge branch 'main' into security/eql-suppression

 Conflicts:
	x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_request_schema.test.ts
	x-pack/plugins/security_solution/common/detection_engine/utils.test.ts
	x-pack/plugins/security_solution/common/experimental_features.ts
	x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/description_step/index.test.tsx
	x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/schema.tsx
	x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/use_experimental_feature_fields_transform.ts
	x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.test.tsx
	x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/use_alert_suppression.tsx
	x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/bulk_create_suppressed_alerts_in_memory.ts
	x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/wrap_suppressed_alerts.ts
	x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts
	x-pack/test/security_solution_cypress/serverless_config.ts

rylnd committed:
Remove duplicated helper

* Replaces `isAlertSuppressionActive` with the (already in main)
  implementation of `getIsAlertSuppressionActive`
* Converts unit tests for `isAlertSuppresionActive` to
  `getIsAlertSuppressionActive` unit tests

rylnd committed:
Fix useAlertSuppression tests

Now that we have two calls to `useIsExperimentalFeatureEnabled`, our
mock needs to be updated.

rylnd commented:
/ci

rylnd committed:
Merge branch 'main' into security/eql-suppression

rylnd commented:
/ci

rylnd committed:
Merge branch 'main' into security/eql-suppression

 Conflicts:
	x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/eql/eql.ts

rylnd committed:
Add missing argument to suppression function in eql executor

This was missed in a previous merge of #178294, but was somehow not
caught by CI. Suspicious!

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

rylnd committed:
Merge branch 'main' into security/eql-suppression

rylnd committed:
Type fixes

The executor version of experimentalFeatures is optional, but we need a
guaranteed version (which comes from createOptions)

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.test.tsx:
```diff
@@ -0,0 +1,165 @@
+
+      jest
+        .spyOn(useAlertSuppressionMock, 'useAlertSuppression')
+        .mockReturnValueOnce({ isSuppressionEnabled: false });
```
has to have `true` here, as test title implies it won't be rendered because `alert_suppression` props missing.
But previous test already asserts no suppression fields will be rendered when `isSuppressionEnabled=false`

rylnd replied:
11077bde861

vitaliidm commented on x-pack/plugins/security_solution/public/detections/components/alerts_table/actions.tsx:
```diff
@@ -274,6 +273,13 @@ export const isThresholdAlert = (ecsData: Ecs): boolean => {
+export const isEqlAlert = (ecsData: Ecs): boolean => {
+  const ruleType = getField(ecsData, ALERT_RULE_TYPE);
+  return (
+    ruleType === 'eql' || (Array.isArray(ruleType) && ruleType.length > 0 && ruleType[0] === 'eql')
```
```suggestion
    ruleType === 'eql' || (Array.isArray(ruleType) && ruleType?.[0] === 'eql')
```

rylnd replied:
67a0b6d6452

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/eql/eql.ts:
```diff
@@ -110,10 +126,36 @@ export const eqlExecutor = async ({
+          newSignals = wrapHits(events, buildReasonMessageForEqlAlert);
+        }
+      } else if (sequences) {
+        if (isAlertSuppressionActive && completeRule.ruleParams.alertSuppression) {
```
I don't think `completeRule.ruleParams.alertSuppression` is needed here.

`isAlertSuppressionActive` is computed before and uses alertSuppression config for this

rylnd replied:
3f731230b4e

vitaliidm commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule_suppression_sequence.cy.ts:
```diff
@@ -0,0 +1,52 @@
+/*
```
Nit: I think test cases in `event_correlation_rule_suppression_sequence.cy.ts` and `event_correlation_rule_suppression.cy.ts` can be combined in one file - less maintanence

rylnd replied:
In the course of reviewing the tests I came to the same conclusion 👍

vitaliidm reviewed (commented):
Thanks for finalising this PR.

I haven't yet tested it or looked at FTR tests.

It's the first quick pass. It looks good.

Can you please this checklist https://docs.elastic.dev/security-solution/process/detections/pull-requests#definition-of-done to PR description, similarly as in new terms PR.
I am interested mostly in flaky test runner 

Will continue with testing and review

rylnd committed:
Fix typo in test filename

rylnd committed:
Refactor our EQL suppression API tests

* Breaks down broadly into sequence/non-sequence, and per suppression
  duration type
* Updates test descriptions to be more accurate/descriptive

rylnd committed:
Normalize EQL cypress tests

* Similiar test descriptions
* Similar formatting

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -446,37 +462,65 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
+    isEqlRule(ruleType) &&
+    isEqlSequenceQuery(queryBar?.query?.query as string) &&
+    groupByFields.length === 0;
+
```
When trying to clear invalid group by fields controller error disappears, but not configured fields.

Upon saving the rule, error appears again

https://github.com/elastic/kibana/assets/92328789/aa221189-51fa-4936-b836-0fe1fa44d842

vitaliidm replied:
does not work when I tried to remove fields one by one

https://github.com/elastic/kibana/assets/92328789/c2fad415-7fa2-4382-9a6e-666dd3fc7835

vitaliidm replied:
Happens on rule editing, not rule creation

rylnd replied:
Can you please provide steps to reproduce this? This appears to be working differently for me: both the clear and individual removal of fields works (on create and edit), and once all fields are removed the suppression fields become disabled.

rylnd replied:
So, cypress tests started failing in 7aa52b8bde8a829ae8e2cedd357b051ce4549767, and when I reverted that to attempt to fix tests, I was able to reproduce the behavior above. So one implementation fails in cypress, and the other causes a UI bug.... 🤔

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -446,37 +462,65 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
    *  - when suppression by rule execution is selected in radio button
-   *  - whe threshold suppression is not enabled and no group by fields selected
+   *  - when threshold suppression is not enabled and no group by fields selected
+   * -  Eql sequence is used and suppression fields are in the default state
```
what does default state mean in this context?

rylnd replied:
There was previously logic in the rule form to keep the suppression fields enabled (but invalid) until they were changed back to their default values. I removed this as part of the form refactor, but this might also be the cause of the UI oddities you observed, so I'll investigate.

rylnd replied:
Update: see https://github.com/elastic/kibana/pull/176422#discussion_r1563610751

vitaliidm reviewed (commented):
I did testing and review ftr tests. Overall it looks good

There are couple of comments I left: on test case and issues on rule create/edit form related to sequence query

rylnd commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql_alert_suppression.ts:
```diff
@@ -0,0 +1,1563 @@
+          });
+        });
+
+        it('suppress alerts with missing fields when rule is configured to do so', async () => {
```
Hmm, I don't think this correct.

rylnd committed:
Replace ConditionalToopTip with EuiToolTip

The conditional version wasn't adding much value here.

rylnd committed:
Fix test subjects and cypress constants for accuracy

These pertain to the "missing fields strategy" field.

rylnd committed:
More accurate test descriptions

rylnd committed:
Fix logical error in test

We're asserting that suppression fields aren't shown because there's
nothing relevant on the rule, not because this hook is disabled (which
is the previous test).

rylnd committed:
Simplify predicate logic

And leverage existing helper to get rid of magic strings.

rylnd committed:
Simplify logic

`isAlertSuppressionActive` already accounts for the presence of these rule
params.

rylnd commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -158,6 +160,20 @@ const IntendedRuleTypeEuiFormRow = styled(RuleTypeEuiFormRow)`
   ${({ theme }) => `padding-left: ${theme.eui.euiSizeXL};`}
 `;
 
+const ConditionalToolTip: FC<{
```
7aa52b8bde8a829ae8e2cedd357b051ce4549767

rylnd commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/rule_definition_section.test.tsx:
```diff
@@ -0,0 +1,165 @@
+      );
+    });
+
+    test('should render only AlertSuppressionTitle and SuppressAlertsDuration when rule type does not have group_by field like threshold', () => {
```
cc95859ecd74482d29326d195a3c1df4aadcf1e8

rylnd committed:
Merge branch 'main' into security/eql-suppression

rylnd committed:
Add integration tests for risk score enrichment and exceptions

rylnd commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/execution_logic/eql_alert_suppression.ts:
```diff
@@ -0,0 +1,1563 @@
+      });
+    });
+  });
+};
```
d85835313f4cfc11fcf5b12628065bd8016c19c2

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

rylnd committed:
Repurpose existing tooltip to display eql suppression tooltip

We already have a tooltip to display the "suppression is not available
for this license" message. Rather than introduce a new tooltip (which
has strange/bad interactions in the UI, as evidenced by cypress
failures), we're now reusing that same tooltip functionality to display
the message about suppression being disabled for sequences.

vitaliidm reviewed (approved):
@rylnd 

Thank you for addressing my comments.

I am approving PR.

Just 2 pending questions to address:
1. Preview enabled when form is invalid https://github.com/elastic/kibana/pull/176422#discussion_r1562274044
2. Discussed earlier issue with tooltips https://github.com/elastic/kibana/pull/176422#discussion_r1562340936

rylnd committed:
Merge branch 'main' into security/eql-suppression

 Conflicts:
	x-pack/test/security_solution_cypress/config.ts

mikecote reviewed (approved):
As mentioned to @rylnd over Slack, starting next week if the intermediate release RFC is approved, we'll want to ensure rule parameter changes like this one are made in an intermediate release manner. I'll approve the PR for now given it's not introducing a new required param.

rylnd commented:
@vitaliidm responding to the outstanding issues here:

> [Preview enabled when form is invalid](https://github.com/elastic/kibana/pull/176422#discussion_r1562274044)

I was expecting the form's invalidity to disable the rule preview, but TIL we have independent logic for determining if the rule preview is disabled. I did [investigate what this functionality would take](https://github.com/elastic/kibana/commit/9b68368ec287966e1d428f5e288ff7db72b2a99d), but ultimately I don't think this is worth doing:

1. This code would be temporary, and removed once EQL sequence suppression is added
2. Between the error being rendered on the form, the inability to save/proceed with the form, and the warning presented on the rule preview (as shown [in your screenshot](https://github.com/elastic/kibana/pull/176422#discussion_r1562274044)), there's plenty of feedback about the invalid state
3. There's no negative impact to previewing the rule in that state (since we already added the guards to the executor, where the warning is shown)
4. The preview/warning is arguably better feedback/UX than a disabled rule preview.

cc @yctercero 

> [Discussed earlier issue with tooltips](https://github.com/elastic/kibana/pull/176422#discussion_r1562340936)

I believe this issue is resolved, now that we've gone back to using a single, non-nested `EuiToolTip`. Can you please verify  this, or provide steps to reproduce?

banderror committed:
Merge branch 'main' into security/eql-suppression

vitaliidm commented:
@rylnd 

> https://github.com/elastic/kibana/pull/176422#discussion_r1562340936
> I believe this issue is resolved, now that we've gone back to using a single, non-nested EuiToolTip. Can you please verify this, or provide steps to reproduce?

Can't reproduce it anymore. Thanks for fixing it.


> 1. This code would be temporary, and removed once EQL sequence suppression is added
> 2. Between the error being rendered on the form, the inability to save/proceed with the form, and the warning presented on the rule preview (as shown https://github.com/elastic/kibana/pull/176422#discussion_r1562274044), there's plenty of feedback about the invalid state
> 3. There's no negative impact to previewing the rule in that state (since we already added the guards to the executor, where the warning is shown)
> 4. The preview/warning is arguably better feedback/UX than a disabled rule preview.


I agree on point 1, it's temporary, so we might just leave it there as it is.
As for the rest, I don't think it gives any new information to user, since warning is essentially the same as error on form.
That rule preview results would be completely irrelevant to the actual rule run, since user won't be able to save that invalid configuration. If they need sequence EQL, they should then remove suppression options. 

I'll leave it with you to decide how do you want to proceed here.

banderror commented on x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_request_schema.mock.ts:
```diff
@@ -213,3 +214,14 @@ export const getUpdateNewTermsSchemaMock = (
 });
+
+export const getCreateEqlRuleSchemaMock = (ruleId = 'rule-1'): EqlRuleCreateProps => ({
+  description: 'Event correlation index pattern rule',
+  name: 'Event correlation index pattern rule',
+  severity: 'high',
+  risk_score: 55,
+  rule_id: ruleId,
+  type: 'eql',
+  language: 'eql',
+  query: 'process where process.name == "regsvr32.exe"',
```
Nit: It's a bit confusing that it's called "index pattern rule" but no index patterns are actually specified.

rylnd replied:
188a053e010

banderror commented on x-pack/plugins/security_solution/common/detection_engine/utils.ts:
```diff
@@ -60,6 +60,9 @@ export const normalizeThresholdField = (
       [thresholdField!];
 };
 
+export const isEqlSequenceQuery = (ruleQuery: string | undefined): boolean =>
+  ruleQuery?.trim().startsWith('sequence') ?? false;
```
Correct me if I'm wrong, but it looks like this check is done only on the FE side. Do we have any additional guards/validation on the BE side against creating a sequence EQL rule with suppression via the API? If we don't, what happens if such a rule is created and started?

UPD: Oh I see, in this case the rule will run without suppression and log a warning status 👍 

```ts
result.warningMessages.push(
  'Suppression is not supported for EQL sequence queries. The rule will proceed without suppression.'
);
```

I think it would be worth mentioning in https://github.com/elastic/security-docs/issues/4977

nastasha-solomon replied:
Hey @rylnd - how is this warning surfaced? Is it returned in an API response when users create and enable an EQL rule via the [create rule endpoint](https://www.elastic.co/guide/en/security/current/rules-api-create.html#_response_payload)? Does the warning ever appear the Security app UI —  and if so, where? For instance, does it show up on the **Execution results** [tab](https://security-docs_bk_5057.docs-preview.app.elstc.co/guide/en/security/master/alerts-ui-monitor.html#rule-execution-logs) within a rule's details page? Or does it appear within the rule creation form when users are setting up a new EQL rule that's using a sequence query? I'm mainly trying to understand whether this warning will be immediately visible to users or if they'll need to seek it out.

banderror reviewed (approved):
Rule Management codeowners changes LGTM 👍 

Reviewed the corresponding diffs, but didn't test the PR -- relying on the owners of the feature here.

Really appreciate all the details in the PR description -- it was helpful to get some context before reviewing the change. Thank you to whoever spent their time on this!

@rylnd Just curious what's the strategy for releasing this feature. Since there's a dedicated feature flag, does that mean that after merging this PR it's not going to go live in Serverless next week, and acceptance/exploratory testing will be done + writing the docs, all after the FF?

banderror committed:
Merge branch 'main' into security/eql-suppression

MadameSheema commented on x-pack/test/security_solution_cypress/config.ts:
```diff
@@ -46,6 +46,7 @@ export default async function ({ readConfigFile }: FtrConfigProviderContext) {
         '--xpack.ruleRegistry.unsafe.legacyMultiTenancy.enabled=true',
         `--xpack.securitySolution.enableExperimental=${JSON.stringify([
           'alertSuppressionForNewTermsRuleEnabled',
+          'alertSuppressionForNonSequenceEqlRuleEnabled',
```
As the FF has been activated here, all the tests are going to be affected by it, not only the ones where the FF has specifically added at the top of the spec file.

rylnd replied:
As far as I'm aware, this is still the only way to run a cypress test locally with a feature flag enabled (https://github.com/elastic/security-team/issues/7423). Please correct me if that's changed.

The global scope is unfortunate, but we don't really have an alternative. If the flag is not enabled globally, then a developer attempting to run a FF test would see it pass on CI and then fail locally; how would they know to set the global config in order to see the test pass?

MadameSheema replied:
You are completely right. 

If I'm not mistaken that behaviour should be documented so everybody should be aware about it. I understand your point, if you prefer to have it globally then you can remove the FF from the spec files.

My comment was more a reminder about what the behaviour is going to be :)

rylnd replied:
>  if you prefer to have it globally

I would certainly prefer _not_ to set the global configuration, as it's both onerous and broad. But until https://github.com/elastic/security-team/issues/7423 is addressed, it's the only way to develop cypress tests requiring a feature flag.

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule_suppression.cy.ts:
```diff
@@ -0,0 +1,137 @@
+describe(
+  'Detection Rule Creation - EQL Rules - With Alert Suppression',
+  {
+    tags: ['@ess', '@serverless'],
```
This test should be marked as `@serverless`. Currently, there is no mechanism to enable through configuration FF on MKI environments, so this test is going to start failing. cc @yctercero

rylnd replied:
You mean `@skipInServerless`, right?

rylnd replied:
66008362aea

yctercero replied:
If we can also leave a note or issue reference when we use the skip tags to denote if there's a bug or issue that needs follow up or if it's an intentional skip that does not require any follow up.

rylnd replied:
Yes, our squashing strategy has unfortunately severely reduced the utility of commit messages; the only way to pass along context is through declarative code or code comments :(

MadameSheema commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule_suppression_sequence.cy.ts:
```diff
@@ -0,0 +1,57 @@
+describe(
+  'Detection Rule Creation - EQL Rules - With Alert Suppression',
+  {
+    tags: ['@ess', '@serverless'],
```
Same as: https://github.com/elastic/kibana/pull/176422/files#r1567339580

rylnd replied:
https://github.com/elastic/kibana/commit/66008362aeab7e3448ea4f8282abaa79a0cbad6b

MadameSheema commented on x-pack/test/security_solution_cypress/serverless_config.ts:
```diff
@@ -35,6 +35,7 @@ export default async function ({ readConfigFile }: FtrConfigProviderContext) {
           { product_line: 'cloud', product_tier: 'complete' },
         ])}`,
         `--xpack.securitySolution.enableExperimental=${JSON.stringify([
+          'alertSuppressionForNonSequenceEqlRuleEnabled',
```
Same as: https://github.com/elastic/kibana/pull/176422/files#r1567337507

logeekal commented:
I did the desk testing. Functionality seems to be working great. Although, I had one question.


In alert table, we can see one icon for suppressed alerts. But I cannot see that icon in timeline. Is this expected? Please see below video for the demo.


https://github.com/elastic/kibana/assets/7485038/dbedb775-a476-4cba-b8e9-f9db9043f57e

rylnd commented:
@logeekal the behavior you've described looks consistent across all rule types supporting suppression (and was not introduced specifically in this PR). I've created an issue to look into this: https://github.com/elastic/kibana/issues/180976.

PhilippeOberti reviewed (approved):
approving for the Threat Hunting Investigations team because @logeekal did the review and his question was answered!

rylnd committed:
Add missing index pattern to rule mock

The name/description indicate that this should contain an index pattern,
but it did not.

rylnd commented:
> Just curious what's the strategy for releasing this feature. Since there's a dedicated feature flag, does that mean that after merging this PR it's not going to go live in Serverless next week, and acceptance/exploratory testing will be done + writing the docs, all after the FF?

@banderror that's mostly correct. We plan merge with the FF disabled, and use the next week or so for testing. Docs are already [in review](https://github.com/elastic/security-docs/pull/5057) so we'll just need to make minor amendments depending on the outcome of testing/release.

At a high level, this PR is relatively low risk as it mainly reuses existing suppression code, and applies it to candidate EQL alerts in a manner nearly identical to several other rule types.

rylnd committed:
Skip cypress tests reliant on feature flag in serverless MKI

Without a mechanism to enable this feature flag in those environments,
these tests will fail and block operations.

rylnd commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_creation/event_correlation_rule_suppression_serverless_essentials.cy.ts:
```diff
@@ -0,0 +1,80 @@
+describe(
+  'Detection Rule Creation - EQL Rules - With Alert Suppression  - Serverless Essentials License',
+  {
+    tags: ['@serverless'],
```
https://github.com/elastic/kibana/commit/66008362aeab7e3448ea4f8282abaa79a0cbad6b

rylnd commented:
@MadameSheema I added the `@skipInServerless` tag for MKI in [6600836](https://github.com/elastic/kibana/pull/176422/commits/66008362aeab7e3448ea4f8282abaa79a0cbad6b); however it occurred to me that a more declarative `@reliesOnFeatureFlag` tag might be a better approach here, since that would both a) explicitly tag the test(s) with that dependency, and then b) allow the relevant CI/Ops pipelines to skip those tests (or not) as they see fit. What do you think?

MadameSheema commented:
> @MadameSheema I added the @skipInServerless tag for MKI in [6600836](https://github.com/elastic/kibana/pull/176422/commits/66008362aeab7e3448ea4f8282abaa79a0cbad6b); however it occurred to me that a more declarative @reliesOnFeatureFlag tag might be a better approach here, since that would both a) explicitly tag the test(s) with that dependency, and then b) allow the relevant CI/Ops pipelines to skip those tests (or not) as they see fit. What do you think

I'm open to add more tags and I personally believe that is a pretty good idea, but there are some objections regarding that.

@yctercero @maximpn thoughts around @rylnd proposal?

MadameSheema reviewed (approved):
sec-eng-prod changes LGTM! Thanks! :)

maximpn commented:
> > @MadameSheema I added the @skipInServerless tag for MKI in [6600836](https://github.com/elastic/kibana/pull/176422/commits/66008362aeab7e3448ea4f8282abaa79a0cbad6b); however it occurred to me that a more declarative @reliesOnFeatureFlag tag might be a better approach here, since that would both a) explicitly tag the test(s) with that dependency, and then b) allow the relevant CI/Ops pipelines to skip those tests (or not) as they see fit. What do you think
> 
> I'm open to add more tags and I personally believe that is a pretty good idea, but there are some objections regarding that.

@MadameSheema AFAIK the majority of feature flags are transient though it's matter of months or years until a feature flag is permanently enabled and removed. If we decide to add `@reliesOnFeatureFlag` flag then someone will need to remove this flag for all related tests.

Additionally `@skipInServerless` and `@reliesOnFeatureFlag` have different meaning. `@skipInServerless` is much broader and it conveys a message in a clear way. The test will be skipped in Serverless (what it exactly means depends on different details but as minimum it's easier to reason about his tag). A test can be skipped when it's broken, flaky or intentionally skipped because it's only a ESS feature or we don't want to run such a test in Serverless. `@reliesOnFeatureFlag` on the other hand doesn't say anything what will happed to the test.

I'd rather leave it to the test to properly set up the environment to enable proper feature flags than adding a new tag. In a case when a test must be skipped it's should be enough to have `@skipInServerless` with a comment.

@yctercero WDYT?

rylnd commented:
@maximpn thanks for your feedback!

>  If we decide to add @reliesOnFeatureFlag flag then someone will need to remove this flag for all related tests.

This is true, and I personally see this behavior as a feature. Consider the alternative: if a test is skipped with `skipInServerless` due to relying on a feature flag, and that feature flag is then removed, how would you go about finding and unskipping that test? At some point, we need to associate "relies on feature flag" and "skipped in serverless," or else we lose that context until someone takes the time to understand the intention/requirements of that test: that it relied on a feature flag, and that it can now be unskipped.

We can do that with a comment, or a tag; I don't really care _how_ we do it. But seeing `skipInServerless` does not in itself tell me _why_ a test is skipped, and I need that information if the test is ever going to be unskipped.

> @skipInServerless is much broader and it conveys a message in a clear way. 

No disagreement here, but it's that same broadness/lack of specificity that I have issue with.

> I'd rather leave it to the test to properly set up the environment to enable proper feature flags

I agree, that's the correct solution. What we're trying to address are symptoms caused by inadequate tooling.

>

maximpn commented:
Hi @rylnd,

I see your point. Do you mind creating a discussion ticket? I think we could assemble all the information there and have more opinions in a centralised way.

rylnd commented:
@maximpn [done](https://github.com/elastic/security-team/issues/9280)!

rylnd commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/step_define_rule/index.tsx:
```diff
@@ -446,37 +462,65 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
   );
 
+  const [{ queryBar }] = useFormData<DefineStepRule>({ form, watch: ['queryBar'] });
+  const areSuppressionFieldsDisabledBySequence =
```
https://github.com/elastic/kibana/issues/182307
